recipes/full_finetune_distributed.py

```diff
@@ -314,41 +289,51 @@ def _setup_model(
                 ac_option,
             )
 
-        # Wrap the model with FSDP. This will ensure that the model is sharded
-        # across all available GPUs.
-        model = FSDP(
-            module=model,
-            auto_wrap_policy=utils.get_full_finetune_fsdp_wrap_policy(
-                memory_efficient_fsdp_wrap=memory_efficient_fsdp_wrap,
-                modules_to_wrap={modules.TransformerDecoderLayer},
-            ),
-            cpu_offload=CPUOffload(offload_params=fsdp_cpu_offload),
-            sharding_strategy=torch.distributed.fsdp.ShardingStrategy.FULL_SHARD,
-            device_id=self._device,
-            # this recipe does not currently support mixed precision training
-            mixed_precision=None,
-            # Ensure we broadcast params and buffers from rank 0
-            sync_module_states=True,
-            # Initialize empty modules on all non-zero ranks
-            param_init_fn=(
-                lambda module: module.to_empty(
-                    device=torch.device("cuda"), recurse=False
-                )
-                if not self._is_rank_zero
-                else None
-            ),
-        )
-
-        # Ensure no params and buffers are on meta device
-        utils.validate_no_params_on_meta_device(model)
-
-        # original activation checkpointing (full) - flip the condition above
         if enable_activation_checkpointing and ac_mode is None:
             utils.set_activation_checkpointing(
                 model, auto_wrap_policy={modules.TransformerDecoderLayer}
             )
 
+        fsdp_kwargs = {}
+        if fsdp_cpu_offload:
+            fsdp_kwargs["offload_policy"] = CPUOffloadPolicy()
```

Comment:
should we expose the shard strategy like in https://github.com/pytorch/torchtune/pull/1024

Reply:
We can, but imo at this stage it's a nice-to-have rather than a must-have

Reply:
It does feel wrong to remove a feature that was just recently contributed though. It feels the goal of moving this out of dev is to integrate it with everything we already have.

Reply:
It hasn't landed yet though

Reply:
To follow up on this, there is not a 1:1 mapping between FSDP1 sharding strategy and FSDP2 APIs. But 

(1) for single-node training (all we currently support), the only relevant sharding strategies should be FULL_SHARD, SHARD_GRAD_OP, and NO_SHARD. HYBRID_SHARD and _HYBRID_SHARD_ZERO2 are only distinct from these in the case of multi-node training.
(2) FSDP2 does not support NO_SHARD, in that case they just redirect users to DDP

So then we are just down to FULL_SHARD and SHARD_GRAD_OP, which correspond to setting [reshard_after_forward](https://github.com/pytorch/pytorch/blob/345bea01dc69d9c31a4954c299767ccecd702d3d/torch/distributed/_composable/fsdp/fully_shard.py#L33) to True and False, respectively. So I can expose this flag in the recipe.

Thanks to @weifengpy for helping to clear this up!